controller/src/main/java/io/pravega/controller/store/stream/ZKGarbageCollector.java

```diff
@@ -0,0 +1,241 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.controller.store.stream;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.util.concurrent.AbstractService;
+import io.pravega.common.concurrent.Futures;
+import io.pravega.common.util.BitConverter;
+import io.pravega.controller.store.stream.tables.Data;
+import lombok.AccessLevel;
+import lombok.Getter;
+import lombok.SneakyThrows;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.curator.framework.recipes.cache.NodeCache;
+import org.apache.curator.framework.recipes.cache.NodeCacheListener;
+import org.apache.curator.framework.recipes.leader.LeaderLatch;
+import org.apache.curator.framework.recipes.leader.LeaderLatchListener;
+
+import java.time.Duration;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+
+/**
+ * Garbage Collector class which periodically executes garbage collection over some data in zookeeper.
+ * This is a generic class which takes a garbage collection Identifier (gcName) and a garbage collection lambda and periodically
+ * executes the lambda.
+ * This also implements the Batching logic and the leader keeps generating new long based batch ids in
+ * each GC cycle. It is responsibility of the user to make use of the `batch-id` to group its data into these batches.
+ * Upon each periodic cycle, this class checks if a new batch should be generated, and then it executes the supplied gc lambda.
+ *
+ * All controller instances compete to become leader for GC workflow. The leader is responsible for maintaining the batch id
```

Comment:
As discussed offline, we might want to rely on a CAS operation for this rather than leader election. It would simplify the logic here and leader election is not strictly necessary.

Reply:
Done